File: elide-gradle-plugin/src/main/java/dev/elide/gradle/ElideGradlePlugin.java
package dev.elide.gradle;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.tasks.compile.JavaCompile;

import javax.inject.Inject;
import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;

public class ElideGradlePlugin implements Plugin<Project> {
    // Plugin ID for Gradle's built-in Java support.
    private static final String javaPluginId = "java";

    // Binary name for Elide.
    private static final String elideBinName = "elide";

    // Extension name for configuring Elide's Gradle plugin.
    private static final String elideExtensionName = elideBinName;

    // Project where the plugin is installed.
    private final Project activeProject;

    @Inject
    public ElideGradlePlugin(Project project) {
        this.activeProject = project;
    }

    // Configure a Java compile task to use Elide instead of the standard compiler API.
    private Task configureJavaCompileToUseElide(Project project, JavaCompile task, ElideExtension ext) {
        project.getLogger().info(
                "Installing Elide's javac support for task '{}' within project '{}'",
                task.getName(),
                activeProject.getName());

        var javaHomeShim = Paths.get(System.getProperty("java.home"))
                .resolve("bin")
                .resolve("elide-javac");

        Path resolvedElide = null;
        if (!Files.exists(javaHomeShim)) {
            if (Files.isWritable(javaHomeShim.getParent())) {
                // we can create the shim; if we are configured to do, we should do so now.
                try(var writer = Files.newBufferedWriter(javaHomeShim)) {
                    // write the shim to the file
                    writer.write("#!/bin/sh\n");
                    writer.write("exec " + resolvePathToElide().toAbsolutePath() + " javac -- \"$@\"\n");
                } catch (IOException e) {
                    throw new RuntimeException("Failed to write Elide javac shim", e);
                }
            } else {
                // we can't write the shim, and it's not there, and we need it, so we should warn and fall back.
                project.getLogger().warn("Elide's javac shim was not found at '{}'; falling back to stock javac.",
                        javaHomeShim.toAbsolutePath());
                return task;
            }
        } else if (!Files.isExecutable(javaHomeShim)) {
            // the shim is there, but it's not executable.
            var result = javaHomeShim.toFile().setExecutable(true);
            if (result) {
                // we're good to go
                resolvedElide = javaHomeShim;
            } else {
                // we can't write the shim, and it's not there, and we need it, so we should warn and fall back.
                project.getLogger().warn("Elide's javac shim isn't executable, and can't be made executable Please " +
                                "run 'chmod +x {}' to fix this.",
                        javaHomeShim.toAbsolutePath());
                return task;
            }
        } else {
            // the shim is there and executable, so we can use it.
            resolvedElide = javaHomeShim;
        }
        if (resolvedElide == null) {
            project.getLogger().error("Failed to resolve Elide javac shim, and Java Home is not writable.");
            throw new RuntimeException("Failed to resolve Elide javac shim; is your Java Home writable?");
        }
        Objects.requireNonNull(resolvedElide);
        var pathAsString = resolvedElide.toString();
        var options = task.getOptions();
        var forkOptions = options.getForkOptions();
        options.setFork(true);
        forkOptions.setExecutable(pathAsString);
        return task;
    }

    // Install integration with Gradle's Java plugin; this prefers Elide's Java Compiler support.
    private Collection<Task> installJavacSupport(Path path, Project project) {
        var elideExtension = project.getExtensions().getByType(ElideExtension.class);
        var compileTasks = project.getTasks().withType(JavaCompile.class);
        project.getLogger().info(
                "Installing Elide's javac support for {} Gradle tasks (path: '{}')",
                compileTasks.size(),
                path.toString());

        return compileTasks.stream()
                .map(compileTask -> configureJavaCompileToUseElide(project, compileTask, elideExtension))
                .collect(Collectors.toList());
    }

    // Install integration with Gradle's Maven root support.
    private Collection<Task> installMavenDepsSupport(Project project, ElideExtension ext, boolean generateManifest) {
        var repos = project.getRepositories();
        var localDepsPath = ext.resolveLocalDepsPath();
        repos.mavenLocal(it -> {
            it.setName("elide");
            it.setUrl(URI.create("file://" + localDepsPath.toAbsolutePath()));
        });
        return Collections.emptyList();
    }

    // Resolve the path to use when invoking the Elide binary.
    private Path resolvePathToElide() {
        var path = System.getenv("PATH");
        var pathSplit = path.split(File.pathSeparator);
        var elideViaPath = Arrays.stream(pathSplit)
                .map(Paths::get)
                .map(p -> p.resolve(elideBinName))
                .filter(Files::exists)
                .filter(Files::isExecutable)
                .findFirst();

        // prefer elide on the user's path
        if (elideViaPath.isPresent()) {
            return elideViaPath.get().toAbsolutePath();
        }

        // try the user's home?
        var elideWithinHome = Paths.get(System.getProperty("user.home"))
                .resolve("elide")
                .resolve(elideBinName);
        if (Files.exists(elideWithinHome) && Files.isExecutable(elideWithinHome)) {
            return elideWithinHome.toAbsolutePath();
        }

        throw new RuntimeException("Failed to find `elide` on your PATH; is it installed?");

        // otherwise, we should resolve from the root project's layout. the plugin will download elide and install it
        // in `<rootProject>/build/elide-runtime`.
        //
        // var elideBuildRoot = activeProject.getRootProject().getLayout().getBuildDirectory().dir("elide-runtime");
        // return elideBuildRoot.get().dir("bin").file(elideBinName).getAsFile().toPath();
    }

    // Determine whether Elide's Maven installer integration is enabled.
    private boolean enableMavenInstaller(Project project, ElideExtension ext) {
        var disableInstallerProp = project.findProperty("elide.builder.maven.install.enable");
        if (disableInstallerProp != null) {
            return Boolean.parseBoolean(disableInstallerProp.toString());
        }
        return ext.getEnableInstall().get() && ext.getEnableMavenIntegration().get();
    }

    // Detect any extant or configured project manifest file. If one is present, `elide install` is run unconditionally.
    private boolean detectProjectManifest(Project project, ElideExtension ext) {
        return (
          (ext.getManifest().isPresent() && ext.getManifest().getAsFile().get().exists()) ||
          project.getLayout().getProjectDirectory().file("elide.pkl").getAsFile().exists()
        );
    }

    // Determine whether Elide's javac shim is enabled.
    private boolean enableJavacShim(Project project, ElideExtension ext) {
        var disableShimProp = project.findProperty("elide.builder.javac.enable");
        if (disableShimProp != null) {
            return Boolean.parseBoolean(disableShimProp.toString());
        }
        return ext.getEnableJavaCompiler().get();
    }

    // Call Elide in a subprocess at the provided path, and with the provided args; capture output and return it as a
    // string to the caller.
    private String callElideCaptured(Path path, String[] args) {
        var allArgs = new String[args.length + 1];
        allArgs[0] = path.toAbsolutePath().toString();
        var i = 1;
        for (var arg : args) {
            allArgs[i] = arg;
            i += 1;
        }

        var subproc = new ProcessBuilder().command(allArgs);
        try {
            var proc = subproc.start();
            var builder = new StringBuilder();

            try (BufferedReader br = new BufferedReader(new InputStreamReader(proc.getInputStream()))) {
                String line;
                while ((line = br.readLine()) != null) {
                    builder.append(line).append(System.lineSeparator());
                }
            }
            try (BufferedReader br = new BufferedReader(new InputStreamReader(proc.getErrorStream()))) {
                String line;
                while ((line = br.readLine()) != null) {
                    builder.append(line).append(System.lineSeparator());
                }
            }
            var exit = proc.waitFor();
            if (exit != 0) {
                throw new RuntimeException("Elide failed with exit code " + exit);
            }
            return builder.toString().trim();
        } catch (InterruptedException ixr) {
            throw new RuntimeException("Failed to wait for Elide process", ixr);
        } catch (IOException ioe) {
            throw new RuntimeException("Failed to start Elide captured process", ioe);
        }
    }

    @SuppressWarnings({"deprecation", "UnstableApiUsage"})
    public void apply(Project project) {
        var elideResolved = resolvePathToElide();
        project.getLogger().debug("Elide resolved to '{}'", elideResolved);
        if (!project.getGradle().getStartParameter().isConfigurationCacheRequested()) {
            var versionPrinted = callElideCaptured(elideResolved, new String[]{"--version"});
            var version = versionPrinted.replace("\n", "");
            project.getLogger().lifecycle("Using Elide " + version);
        }

        var objectUtil = project.getObjects();
        var extension = new ElideExtension(project, objectUtil);
        project.getExtensions().add(elideExtensionName, extension);

        project.afterEvaluate(_ -> {
            var pluginManager = project.getPluginManager();
            var javaPluginActive = pluginManager.hasPlugin(javaPluginId);
            var javacSupportActive = enableJavacShim(project, extension);
            var mavenInstallerActive = enableMavenInstaller(project, extension);
            var hasProjectManifest = detectProjectManifest(project, extension);

            project.getLogger().info(
                    "Elide Java support: (pluginActive={}, javacSupport={})",
                    javaPluginActive,
                    javacSupportActive);

            Collection<Task> javacTasks = null;
            if (javaPluginActive && javacSupportActive) {
                javacTasks = installJavacSupport(elideResolved, project);
            }

            boolean mustGenerateManifest = false;  // @TODO implement
            boolean installerEnabled = extension.getEnableInstall().get();
            boolean shouldRunInstallWithOrWithoutMaven = installerEnabled;
            LinkedList<Task> allPrepTasks = new LinkedList<>();

            if (mavenInstallerActive) {
                // to enable integration with Maven dependency installation, we need to inject a local dependency root
                // path, and we need to run `elide install` before compilation runs. dependencies must also be gathered
                // if there is no project manifest file to read from. this comes first so that Gradle is configured to
                // be aware of our Maven dependencies before `elide install` is run.
                allPrepTasks.addAll(installMavenDepsSupport(project, extension, mustGenerateManifest));
                installerEnabled = true;
            }
            if (installerEnabled && hasProjectManifest) {
                // if a project manifest is present, we should run `elide install` regardless of other criteria, as it
                // may install non-JVM dependencies on the user's behalf.
                shouldRunInstallWithOrWithoutMaven = true;
            }
            if (shouldRunInstallWithOrWithoutMaven) {
                // add a precursor task to run `elide install`.
                allPrepTasks.add(project.getTasks().create(ElideTaskName.ELIDE_TASK_INSTALL, Task.class, task -> {
                    task.setGroup("Elide");
                    task.setDescription("Runs `elide install` to prepare the project for compilation.");
                    task.dependsOn(allPrepTasks.stream().filter(it -> it != task).collect(Collectors.toList()));
                    task.doLast(_ -> {
                        var start = System.currentTimeMillis();
                        project.getLogger().info("Running `elide install`");
                        var result = callElideCaptured(elideResolved, new String[]{"install"});
                        var end = System.currentTimeMillis();
                        project.getLogger().info(result);
                        project.getLogger().lifecycle("`elide install` completed in {}ms", (end - start));
                    });
                }));
            }
            if (!allPrepTasks.isEmpty() && javacTasks != null && !javacTasks.isEmpty()) {
                javacTasks.forEach(javacTask -> {
                    javacTask.dependsOn(allPrepTasks);
                });
            }
        });
    }
}

File: elide-gradle-plugin/src/main/kotlin/dev/elide/gradle/ElideGradlePlugin.kt
package dev.elide.gradle

import dev.elide.gradle.cli.ElideCli
import dev.elide.gradle.configuration.ElideDiagnosticsConfiguration
import dev.elide.gradle.configuration.ElideSettings
import dev.elide.gradle.configuration.binary.ElideBinaryResolutionSource
import dev.elide.gradle.configuration.features.ElideFeaturesConfiguration
import dev.elide.gradle.configuration.features.ElideIntegrationStrategy
import dev.elide.gradle.internal.Platform
import dev.elide.gradle.internal.elideDebug
import dev.elide.gradle.internal.mapNotNull
import dev.elide.gradle.service.ElideThreadPoolService
import dev.elide.gradle.task.ElideCheckVersionTask
import dev.elide.gradle.task.GenerateModuleBuildConfigurationTask
import dev.elide.gradle.task.PrepareElideJavadocShimTask
import dev.elide.gradle.task.download.DownloadElideBinaryTask
import dev.elide.gradle.task.download.ExtractElideBinaryTask
import dev.elide.gradle.task.download.PrepareDownloadedElideBinary
import dev.elide.gradle.task.download.VerifyElideBinaryTask
import dev.elide.gradle.task.exec.ElideCliExec
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.artifacts.ModuleDependency
import org.gradle.api.artifacts.repositories.MavenArtifactRepository
import org.gradle.api.provider.Provider
import org.gradle.api.tasks.TaskProvider
import org.gradle.api.tasks.compile.JavaCompile
import org.gradle.api.tasks.javadoc.Javadoc
import org.gradle.kotlin.dsl.create
import org.gradle.kotlin.dsl.register
import org.gradle.kotlin.dsl.withType
import java.nio.file.Path
import kotlin.io.path.absolutePathString
import kotlin.io.path.exists
import kotlin.io.path.isSameFileAs
import kotlin.io.path.toPath

public class ElideGradlePlugin : Plugin<Project> {
    internal companion object {
        const val THREAD_POOL_SERVICE_NAME = "elideThreadPoolService"
    }

    override fun apply(target: Project) {
        val extension = target.extensions.create<ElideExtension>(
            name = "elide",
            constructionArguments = arrayOf(target)
        )
        val settings = extension.settings

        target.gradle.sharedServices.registerIfAbsent(
            THREAD_POOL_SERVICE_NAME, ElideThreadPoolService::class.java,
        )

        val localCliPathProvider = target.objects.directoryProperty().fileProvider(
            settings.binary.resolutionSource.flatMap {
                ElideCli.resolvePathToCli(target.logger, target.providers)
                    .orElse(
                        settings.binary.resolutionSource.mapNotNull {
                            when (it) {
                                is ElideBinaryResolutionSource.LocalOnly -> null
                                is ElideBinaryResolutionSource.LocalIfApplicable -> it.downloadPath
                                is ElideBinaryResolutionSource.Project -> it.downloadPath
                            }
                        }
                    )
            }.map { it.toFile() }
        )

        val cliDownloadPath = target.objects.directoryProperty().fileProvider(
            target.provider {
                if (settings.binary.resolutionSource.get() is ElideBinaryResolutionSource.LocalOnly)
                    return@provider null

                if (settings.binary.version.orNull == null)
                    error("Elide binary resolution source is set not to be local-only, but version isn't provided.")

                target.layout
                    .buildDirectory
                    .dir("elide-runtime/bin")
                    .map { it.dir("${settings.binary.version.get()}-${Platform.platformClassifier}") }
                    .get()
                    .asFile
            }
        )

        val resultingCliPath = localCliPathProvider.orElse(cliDownloadPath).map {
            it.file(ElideCli.ELIDE_BINARY_NAME).asFile.toPath()
        }

        val elideCli = ElideCli(
            cwd = target.objects.directoryProperty().fileProvider(
                target.layout.buildDirectory.dir("elide-runtime").map { it.asFile }
            ),
            bin = resultingCliPath,
            providerFactory = target.providers,
        )

        val downloadElideCli = target.tasks.register<DownloadElideBinaryTask>("downloadElideCli") {
            targetDirectory.set(cliDownloadPath)
            resolutionSource.set(settings.binary.resolutionSource)
            targetVersion.set(settings.binary.version)
        }

        val verifyBinaryCli = target.tasks.register<VerifyElideBinaryTask>("verifyElideCli") {
            dependsOn(downloadElideCli)

            targetDirectory.set(cliDownloadPath)
        }

        val extractElideCli = target.tasks.register<ExtractElideBinaryTask>("extractElideCli") {
            dependsOn(verifyBinaryCli)

            targetDirectory.set(cliDownloadPath)
        }

        target.tasks.register<PrepareDownloadedElideBinary>("prepareDownloadedElideCli") {
            dependsOn(extractElideCli)

            downloadedElideBinary.set(
                target.objects
                    .fileProperty()
                    .fileProvider(cliDownloadPath.file(ElideCli.ELIDE_BINARY_NAME).map { it.asFile })
            )
        }

        target.tasks.register<ElideCheckVersionTask>("checkElideCliVersion") {
            cli.set(elideCli)
            binPath.set(resultingCliPath.map { it.toRealPath().absolutePathString() })
            resolutionSource.set(extension.settings.binary.resolutionSource)
            strictVersionCheck.set(extension.settings.binary.strictVersionCheck)
            targetVersion.set(extension.settings.binary.version)
            silentMode.set(extension.settings.binary.silentMode)
        }

        target.tasks.register<GenerateModuleBuildConfigurationTask>("generateModuleMetadataForElide") {
            onlyIf { extension.settings.features.generatePklBuildConfiguration.get() }

            // Main dependencies
            // Elide for now can't distinguish implementation, compileOnly or runtimeOnly, so we put it all in one
            mainDeclaredDependencies.set(
                target.provider {
                    listOf("implementation", "compileOnly", "runtimeOnly")
                        .mapNotNull { configName ->
                            target.configurations.findByName(configName)
                        }
                        .flatMap { config ->
                            config.dependencies.withType<ModuleDependency>().mapNotNull { dep ->
                                val group = dep.group ?: return@mapNotNull null
                                val name = dep.name
                                val version = dep.version ?: return@mapNotNull null
                                "$group:$name:$version"
                            }
                        }
                }
            )

            // Test dependencies
            // Elide for now can't distinguish implementation, compileOnly or runtimeOnly, so we put it all in one
            testDeclaredDependencies.set(
                target.provider {
                    listOf("testImplementation", "testCompileOnly", "testRuntimeOnly")
                        .mapNotNull { configName ->
                            target.configurations.findByName(configName)
                        }
                        .flatMap { config ->
                            config.dependencies.withType<ModuleDependency>().mapNotNull { dep ->
                                val group = dep.group ?: return@mapNotNull null
                                val name = dep.name
                                val version = dep.version ?: return@mapNotNull null
                                "$group:$name:$version"
                            }
                        }
                }
            )

            declaredRepositories.set(
                target.provider {
                    target.repositories.filterIsInstance<MavenArtifactRepository>()
                        .filter { repo ->
                            repo.url.scheme != "file" ||
                                repo.url.toPath().isSameFileAs(
                                    extension.settings.devRoot.dir("m2/dependencies").get().asFile.toPath()
                                )
                        }
                }
            )

            generatedFile.set(
                target.layout
                    .buildDirectory
                    .dir("elide-runtime/generated")
                    .map { it.file("module.pkl") }
            )
        }

        val elideInstall = extension.exec("elideInstall") {
            group = "Elide"
            description = "Runs `elide install` to prepare the project for compilation."

            args.add("install")
            args.add(telemetry.mapNotNull { enabled -> if (!enabled) "--no-telemetry" else null })

            onlyIf("`elide.settings.manifest` is not set or target file does not exists") {
                settings.manifest.isPresent && settings.manifest.asFile.get().exists()
            }

            onlyIf("`elide.settings.features.enableElideInstall` is false") {
                settings.features.enableElideInstall.get()
            }
        }

        val prepareElideJavadocShimTask = target.tasks.register<PrepareElideJavadocShimTask>(
            name = "prepareElideJavadocShim"
        ) {
            cliPath.set(resultingCliPath.get().absolutePathString())
            shimFile.set(target.layout.buildDirectory.file("elide-runtime/shim/javadoc"))
        }

        target.applyMavenLocalRepo(settings)
        target.overrideJavaCompileIfEnabled(
            featuresConfig = settings.features,
            diagnosticsConfig = settings.diagnostics,
            cliPath = resultingCliPath,
            elideInstall = elideInstall,
        )
        target.overrideJavadocIfEnabled(
            featuresConfig = settings.features,
            cliPath = resultingCliPath,
            prepareElideJavadocShimTask = prepareElideJavadocShimTask,
            silentMode = settings.binary.silentMode,
            shimPath = prepareElideJavadocShimTask.flatMap { file ->
                file.shimFile.asFile.map { it.absolutePath }
            },
        )

        target.configurations.configureEach {
            incoming.beforeResolve {
                // Let's validate that 'elideInstall' is there
                if (settings.features.enableElideInstall.get() && !target.gradle.taskGraph.hasTask(elideInstall.get())) {
                    error("'elideInstall' should run before any dependency resolution. Did somebody started resolution too early (e.g on configuration phase)?")
                }
            }
        }
    }

    /**
     * Adds the maven local repository pointing to the "[dev.elide.gradle.configuration.ElideSettings.devRoot]/dependencies/m2" folder.
     *
     * This function also reorders the list of maven repositories to ensure that local repository is the first one to be used.
     */
    private fun Project.applyMavenLocalRepo(extension: ElideSettings) {
        val localM2DirPath = extension.devRoot
            .get()
            .dir("dependencies")
            .dir("m2")
            .asFile
            .also { it.mkdirs() }
            .absolutePath

        val indexOfLocalElideRepo = repositories.indexOfFirst {
            it is MavenArtifactRepository
                && it.url.scheme == "file"
                && it.url.path == localM2DirPath
        }.takeIf { it != -1 }

        // early return in case if it's already the first one.
        if (indexOfLocalElideRepo == 0) return

        if (indexOfLocalElideRepo != null) {
            repositories.removeAt(indexOfLocalElideRepo)
        }

        // We need to ensure that our maven local repository is the first one, for not to hit
        // remote ones.
        val snapshot = repositories.toList()
        repositories.clear()
        repositories.mavenLocal {
            uri("file://$localM2DirPath")
        }
        snapshot.forEach(repositories::add)
    }

    private fun Project.overrideJavaCompileIfEnabled(
        featuresConfig: ElideFeaturesConfiguration,
        diagnosticsConfig: ElideDiagnosticsConfiguration,
        cliPath: Provider<Path>,
        elideInstall: TaskProvider<ElideCliExec>,
    ) {
        afterEvaluate {
            val javacStrategy = featuresConfig.javacStrategy.get()

            if (javacStrategy == ElideIntegrationStrategy.NO_ELIDE) {
                logger.elideDebug("Skipping altering JavaCompile tasks: ElideIntegrationStrategy is NO_ELIDE.")
                return@afterEvaluate
            }

            if (cliPath.orNull?.exists() != true && gradle.startParameter.isOffline) {
                logger.elideDebug("Skipping altering JavaCompile tasks: elide is not found and Gradle is in offline mode.")
                return@afterEvaluate
            }

            tasks.withType<JavaCompile>().configureEach {
                dependsOn(elideInstall)

                options.isFork = true
                options.forkOptions {
                    executable = cliPath.get().absolutePathString()
                    jvmArgs = buildList {
                        if (diagnosticsConfig.debug.get())
                            add("--debug")
                        if (diagnosticsConfig.verbose.get())
                            add("--verbose")

                        add("javac")
                        add("--")
                        jvmArgs.orEmpty().forEach {
                            add(it)
                        }
                    }
                }
            }
        }
    }

    private fun Project.overrideJavadocIfEnabled(
        featuresConfig: ElideFeaturesConfiguration,
        cliPath: Provider<Path>,
        shimPath: Provider<String>,
        prepareElideJavadocShimTask: TaskProvider<PrepareElideJavadocShimTask>,
        silentMode: Provider<Boolean>,
    ) {
        afterEvaluate {
            val javadocStrategy = featuresConfig.javadocStrategy.get()

            if (javadocStrategy == ElideIntegrationStrategy.NO_ELIDE) {
                logger.elideDebug("Skipping altering Javadoc tasks: ElideIntegrationStrategy is NO_ELIDE.")
                return@afterEvaluate
            }

            if (cliPath.orNull == null && !silentMode.get()) {
                error("CLI path has not been found. Have you set local-only?")
            }

            if (cliPath.orNull?.exists() != true && gradle.startParameter.isOffline && silentMode.get()) {
                logger.elideDebug("Skipping altering Javadoc tasks: elide is not found and Gradle is in offline mode.")
                return@afterEvaluate
            }

            tasks.withType<Javadoc>().configureEach {
                executable = shimPath.get()
                dependsOn(prepareElideJavadocShimTask)
            }
        }
    }
}
